File: BrickKit/bricks/src/androidTest/java/com/wayfair/brickkit/BrickRecyclerItemDecorationTest.java
/*
 * Copyright © 2017-2020 Wayfair. All rights reserved.
 */
package com.wayfair.brickkit;

import android.content.Context;
import android.graphics.Rect;
import android.view.View;

import com.wayfair.brickkit.brick.BaseBrick;
import com.wayfair.brickkit.padding.BrickPadding;
import com.wayfair.brickkit.size.BrickSize;

import org.junit.Before;
import org.junit.Test;
import org.mockito.Mock;
import org.mockito.MockitoAnnotations;

import java.util.LinkedList;

import androidx.recyclerview.widget.RecyclerView;

import static org.junit.Assert.assertEquals;
import static org.mockito.ArgumentMatchers.nullable;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.when;

public class BrickRecyclerItemDecorationTest {

    private static final int INNER_LEFT = 1;
    private static final int INNER_TOP = 2;
    private static final int INNER_RIGHT = 3;
    private static final int INNER_BOTTOM = 4;
    private static final int OUTER_LEFT = 5;
    private static final int OUTER_TOP = 6;
    private static final int OUTER_RIGHT = 7;
    private static final int OUTER_BOTTOM = 8;
    private static final int SINGLE_COLUMN = BrickDataManager.SPAN_COUNT;
    private static final int MULTI_COLUMN = BrickDataManager.SPAN_COUNT - 1;

    private BrickRecyclerItemDecoration itemDecoration;
    private RecyclerView parent;
    private Rect outRect;
    private View view;
    private RecyclerView.State state;
    @Mock
    private LinkedList<BaseBrick> bricks;
    private BaseBrick brick;
    private BrickSize brickSize;

    @Before
    public void setup() {
        MockitoAnnotations.openMocks(this);
        view = mock(View.class);

        state = mock(RecyclerView.State.class);

        parent = mock(RecyclerView.class);
        when(parent.getChildAdapterPosition(any(View.class))).thenReturn(0);

        BrickDataManager brickDataManager = mock(BrickDataManager.class);
        when(brickDataManager.getRecyclerViewItems()).thenReturn(bricks);

        BrickRecyclerAdapter adapter = mock(BrickRecyclerAdapter.class);
        when(parent.getAdapter()).thenReturn(adapter);
        when(adapter.getBrickDataManager()).thenReturn(brickDataManager);
        itemDecoration = new BrickRecyclerItemDecoration(brickDataManager);

        outRect = new Rect();

        brick = mock(BaseBrick.class);

        BrickPadding brickPadding = mock(BrickPadding.class);
        when(brickPadding.getInnerLeftPadding()).thenReturn(INNER_LEFT);
        when(brickPadding.getInnerTopPadding()).thenReturn(INNER_TOP);
        when(brickPadding.getInnerRightPadding()).thenReturn(INNER_RIGHT);
        when(brickPadding.getInnerBottomPadding()).thenReturn(INNER_BOTTOM);
        when(brickPadding.getOuterLeftPadding()).thenReturn(OUTER_LEFT);
        when(brickPadding.getOuterTopPadding()).thenReturn(OUTER_TOP);
        when(brickPadding.getOuterRightPadding()).thenReturn(OUTER_RIGHT);
        when(brickPadding.getOuterBottomPadding()).thenReturn(OUTER_BOTTOM);
        when(brick.getPadding()).thenReturn(brickPadding);

        brickSize = mock(BrickSize.class);
        when(brick.getSpanSize()).thenReturn(brickSize);

        when(parent.getChildAdapterPosition(any(View.class))).thenReturn(0);
        when(bricks.get(0)).thenReturn(brick);
    }

    @Test
    public void testInvalidPosition() {
        when(parent.getChildAdapterPosition(any(View.class))).thenReturn(-1);

        itemDecoration.getItemOffsets(outRect, view, parent, state);

        assertEquals(0, outRect.left);
        assertEquals(0, outRect.top);
        assertEquals(0, outRect.right);
        assertEquals(0, outRect.bottom);
    }

    @Test
    public void testInvalidNullItem() {
        when(bricks.get(0)).thenReturn(null);

        itemDecoration.getItemOffsets(outRect, view, parent, state);

        assertEquals(0, outRect.left);
        assertEquals(0, outRect.top);
        assertEquals(0, outRect.right);
        assertEquals(0, outRect.bottom);
    }

    @Test
    public void testSingleColumnFirstRow() {
        when(brickSize.getSpans(nullable(Context.class))).thenReturn(SINGLE_COLUMN);
        when(brick.isInFirstRow()).thenReturn(true);
        when(brick.isInLastRow()).thenReturn(false);

        itemDecoration.getItemOffsets(outRect, view, parent, state);

        assertEquals(OUTER_LEFT, outRect.left);
        assertEquals(OUTER_TOP, outRect.top);
        assertEquals(OUTER_RIGHT, outRect.right);
        assertEquals(INNER_BOTTOM, outRect.bottom);
    }

    @Test
    public void testSingleColumnLastRow() {
        when(brickSize.getSpans(nullable(Context.class))).thenReturn(SINGLE_COLUMN);
        when(brick.isInFirstRow()).thenReturn(false);
        when(brick.isInLastRow()).thenReturn(true);

        itemDecoration.getItemOffsets(outRect, view, parent, state);

        assertEquals(OUTER_LEFT, outRect.left);
        assertEquals(INNER_TOP, outRect.top);
        assertEquals(OUTER_RIGHT, outRect.right);
        assertEquals(OUTER_BOTTOM, outRect.bottom);
    }

    @Test
    public void testSingleColumnFirstRowLastRow() {
        when(brickSize.getSpans(nullable(Context.class))).thenReturn(SINGLE_COLUMN);
        when(brick.isInFirstRow()).thenReturn(true);
        when(brick.isInLastRow()).thenReturn(true);

        itemDecoration.getItemOffsets(outRect, view, parent, state);

        assertEquals(OUTER_LEFT, outRect.left);
        assertEquals(OUTER_TOP, outRect.top);
        assertEquals(OUTER_RIGHT, outRect.right);
        assertEquals(OUTER_BOTTOM, outRect.bottom);
    }

    @Test
    public void testSingleColumn() {
        when(brickSize.getSpans(nullable(Context.class))).thenReturn(SINGLE_COLUMN);
        when(brick.isInFirstRow()).thenReturn(false);
        when(brick.isInLastRow()).thenReturn(false);

        itemDecoration.getItemOffsets(outRect, view, parent, state);

        assertEquals(OUTER_LEFT, outRect.left);
        assertEquals(INNER_TOP, outRect.top);
        assertEquals(OUTER_RIGHT, outRect.right);
        assertEquals(INNER_BOTTOM, outRect.bottom);
    }

    @Test
    public void testMultiColumnLeftColumnFirstRow() {
        when(brickSize.getSpans(nullable(Context.class))).thenReturn(MULTI_COLUMN);
        when(brick.isInFirstRow()).thenReturn(true);
        when(brick.isInLastRow()).thenReturn(false);
        when(brick.isOnLeftWall()).thenReturn(true);
        when(brick.isOnRightWall()).thenReturn(false);

        itemDecoration.getItemOffsets(outRect, view, parent, state);

        assertEquals(OUTER_LEFT, outRect.left);
        assertEquals(OUTER_TOP, outRect.top);
        assertEquals(INNER_RIGHT, outRect.right);
        assertEquals(INNER_BOTTOM, outRect.bottom);
    }

    @Test
    public void testMultiColumnMiddleColumnFirstRow() {
        when(brickSize.getSpans(nullable(Context.class))).thenReturn(MULTI_COLUMN);
        when(brick.isInFirstRow()).thenReturn(true);
        when(brick.isInLastRow()).thenReturn(false);
        when(brick.isOnLeftWall()).thenReturn(false);
        when(brick.isOnRightWall()).thenReturn(false);

        itemDecoration.getItemOffsets(outRect, view, parent, state);

        assertEquals(INNER_LEFT, outRect.left);
        assertEquals(OUTER_TOP, outRect.top);
        assertEquals(INNER_RIGHT, outRect.right);
        assertEquals(INNER_BOTTOM, outRect.bottom);
    }

    @Test
    public void testMultiColumnRightColumnFirstRow() {
        when(brickSize.getSpans(nullable(Context.class))).thenReturn(MULTI_COLUMN);
        when(brick.isInFirstRow()).thenReturn(true);
        when(brick.isInLastRow()).thenReturn(false);
        when(brick.isOnLeftWall()).thenReturn(false);
        when(brick.isOnRightWall()).thenReturn(true);

        itemDecoration.getItemOffsets(outRect, view, parent, state);

        assertEquals(INNER_LEFT, outRect.left);
        assertEquals(OUTER_TOP, outRect.top);
        assertEquals(OUTER_RIGHT, outRect.right);
        assertEquals(INNER_BOTTOM, outRect.bottom);
    }

    @Test
    public void testMultiColumnLeftColumnLastRow() {
        when(brickSize.getSpans(nullable(Context.class))).thenReturn(MULTI_COLUMN);
        when(brick.isInFirstRow()).thenReturn(false);
        when(brick.isInLastRow()).thenReturn(true);
        when(brick.isOnLeftWall()).thenReturn(true);
        when(brick.isOnRightWall()).thenReturn(false);

        itemDecoration.getItemOffsets(outRect, view, parent, state);

        assertEquals(OUTER_LEFT, outRect.left);
        assertEquals(INNER_TOP, outRect.top);
        assertEquals(INNER_RIGHT, outRect.right);
        assertEquals(OUTER_BOTTOM, outRect.bottom);
    }

    @Test
    public void testMultiColumnMiddleColumnLastRow() {
        when(brickSize.getSpans(nullable(Context.class))).thenReturn(MULTI_COLUMN);
        when(brick.isInFirstRow()).thenReturn(false);
        when(brick.isInLastRow()).thenReturn(true);
        when(brick.isOnLeftWall()).thenReturn(false);
        when(brick.isOnRightWall()).thenReturn(false);

        itemDecoration.getItemOffsets(outRect, view, parent, state);

        assertEquals(INNER_LEFT, outRect.left);
        assertEquals(INNER_TOP, outRect.top);
        assertEquals(INNER_RIGHT, outRect.right);
        assertEquals(OUTER_BOTTOM, outRect.bottom);
    }

    @Test
    public void testMultiColumnRightColumnLastRow() {
        when(brickSize.getSpans(nullable(Context.class))).thenReturn(MULTI_COLUMN);
        when(brick.isInFirstRow()).thenReturn(false);
        when(brick.isInLastRow()).thenReturn(true);
        when(brick.isOnLeftWall()).thenReturn(false);
        when(brick.isOnRightWall()).thenReturn(true);

        itemDecoration.getItemOffsets(outRect, view, parent, state);

        assertEquals(INNER_LEFT, outRect.left);
        assertEquals(INNER_TOP, outRect.top);
        assertEquals(OUTER_RIGHT, outRect.right);
        assertEquals(OUTER_BOTTOM, outRect.bottom);
    }

    @Test
    public void testMultiColumnLeftColumnFirstRowLastRow() {
        when(brickSize.getSpans(nullable(Context.class))).thenReturn(MULTI_COLUMN);
        when(brick.isInFirstRow()).thenReturn(true);
        when(brick.isInLastRow()).thenReturn(true);
        when(brick.isOnLeftWall()).thenReturn(true);
        when(brick.isOnRightWall()).thenReturn(false);

        itemDecoration.getItemOffsets(outRect, view, parent, state);

        assertEquals(OUTER_LEFT, outRect.left);
        assertEquals(OUTER_TOP, outRect.top);
        assertEquals(INNER_RIGHT, outRect.right);
        assertEquals(OUTER_BOTTOM, outRect.bottom);
    }

    @Test
    public void testMultiColumnMiddleColumnFirstRowLastRow() {
        when(brickSize.getSpans(nullable(Context.class))).thenReturn(MULTI_COLUMN);
        when(brick.isInFirstRow()).thenReturn(true);
        when(brick.isInLastRow()).thenReturn(true);
        when(brick.isOnLeftWall()).thenReturn(false);
        when(brick.isOnRightWall()).thenReturn(false);

        itemDecoration.getItemOffsets(outRect, view, parent, state);

        assertEquals(INNER_LEFT, outRect.left);
        assertEquals(OUTER_TOP, outRect.top);
        assertEquals(INNER_RIGHT, outRect.right);
        assertEquals(OUTER_BOTTOM, outRect.bottom);
    }

    @Test
    public void testMultiColumnRightColumnFirstRowLastRow() {
        when(brickSize.getSpans(nullable(Context.class))).thenReturn(MULTI_COLUMN);
        when(brick.isInFirstRow()).thenReturn(true);
        when(brick.isInLastRow()).thenReturn(true);
        when(brick.isOnLeftWall()).thenReturn(false);
        when(brick.isOnRightWall()).thenReturn(true);

        itemDecoration.getItemOffsets(outRect, view, parent, state);

        assertEquals(INNER_LEFT, outRect.left);
        assertEquals(OUTER_TOP, outRect.top);
        assertEquals(OUTER_RIGHT, outRect.right);
        assertEquals(OUTER_BOTTOM, outRect.bottom);
    }

    @Test
    public void testMultiColumnLeftColumn() {
        when(brickSize.getSpans(nullable(Context.class))).thenReturn(MULTI_COLUMN);
        when(brick.isInFirstRow()).thenReturn(false);
        when(brick.isInLastRow()).thenReturn(false);
        when(brick.isOnLeftWall()).thenReturn(true);
        when(brick.isOnRightWall()).thenReturn(false);

        itemDecoration.getItemOffsets(outRect, view, parent, state);

        assertEquals(OUTER_LEFT, outRect.left);
        assertEquals(INNER_TOP, outRect.top);
        assertEquals(INNER_RIGHT, outRect.right);
        assertEquals(INNER_BOTTOM, outRect.bottom);
    }

    @Test
    public void testMultiColumnMiddleColumn() {
        when(brickSize.getSpans(nullable(Context.class))).thenReturn(MULTI_COLUMN);
        when(brick.isInFirstRow()).thenReturn(false);
        when(brick.isInLastRow()).thenReturn(false);
        when(brick.isOnLeftWall()).thenReturn(false);
        when(brick.isOnRightWall()).thenReturn(false);

        itemDecoration.getItemOffsets(outRect, view, parent, state);

        assertEquals(INNER_LEFT, outRect.left);
        assertEquals(INNER_TOP, outRect.top);
        assertEquals(INNER_RIGHT, outRect.right);
        assertEquals(INNER_BOTTOM, outRect.bottom);
    }

    @Test
    public void testMultiColumnRightColumn() {
        when(brickSize.getSpans(nullable(Context.class))).thenReturn(MULTI_COLUMN);
        when(brick.isInFirstRow()).thenReturn(false);
        when(brick.isInLastRow()).thenReturn(false);
        when(brick.isOnLeftWall()).thenReturn(false);
        when(brick.isOnRightWall()).thenReturn(true);

        itemDecoration.getItemOffsets(outRect, view, parent, state);

        assertEquals(INNER_LEFT, outRect.left);
        assertEquals(INNER_TOP, outRect.top);
        assertEquals(OUTER_RIGHT, outRect.right);
        assertEquals(INNER_BOTTOM, outRect.bottom);
    }
}

File: BrickKit/bricks/src/androidTest/java/com/wayfair/brickkit/BrickRecyclerItemDecorationTest.kt
/*
 * Copyright © 2017-2021 Wayfair. All rights reserved.
 */
package com.wayfair.brickkit

import android.graphics.Rect
import android.view.View
import androidx.recyclerview.widget.RecyclerView
import org.mockito.kotlin.mock
import org.mockito.kotlin.whenever
import com.wayfair.brickkit.brick.BaseBrick
import com.wayfair.brickkit.padding.BrickPadding
import org.junit.Assert.assertEquals
import org.junit.Before
import org.junit.Test

class BrickRecyclerItemDecorationTest {
    private val parent: RecyclerView = mock()
    private val outRect: Rect = Rect()
    private val view: View = mock()
    private val state: RecyclerView.State = mock()
    private val bricks = mutableListOf<BaseBrick>()
    private val brick: BaseBrick = mock()
    private lateinit var itemDecoration: BrickRecyclerItemDecoration

    @Before
    fun setup() {
        whenever(parent.getChildAdapterPosition(view)).thenReturn(0)

        val brickDataManager = mock<BrickDataManager>()
        whenever(brickDataManager.recyclerViewItems).thenReturn(bricks)

        val adapter = mock<BrickRecyclerAdapter>()
        whenever(adapter.brickDataManager).thenReturn(brickDataManager)
        whenever(parent.adapter).thenReturn(adapter)

        itemDecoration = BrickRecyclerItemDecoration(brickDataManager)

        val brickPadding = mock<BrickPadding>()
        whenever(brickPadding.innerLeftPadding).thenReturn(INNER_LEFT)
        whenever(brickPadding.innerTopPadding).thenReturn(INNER_TOP)
        whenever(brickPadding.innerRightPadding).thenReturn(INNER_RIGHT)
        whenever(brickPadding.innerBottomPadding).thenReturn(INNER_BOTTOM)
        whenever(brickPadding.outerLeftPadding).thenReturn(OUTER_LEFT)
        whenever(brickPadding.outerTopPadding).thenReturn(OUTER_TOP)
        whenever(brickPadding.outerRightPadding).thenReturn(OUTER_RIGHT)
        whenever(brickPadding.outerBottomPadding).thenReturn(OUTER_BOTTOM)
        whenever(brick.padding).thenReturn(brickPadding)
        whenever(parent.getChildAdapterPosition(view)).thenReturn(0)
        bricks.add(brick)
    }

    @Test
    fun testInvalidNullItem() {
        bricks.clear()

        itemDecoration.getItemOffsets(outRect, view, parent, state)

        assertEquals(0, outRect.left.toLong())
        assertEquals(0, outRect.top.toLong())
        assertEquals(0, outRect.right.toLong())
        assertEquals(0, outRect.bottom.toLong())
    }

    @Test
    fun testInvalidWrongDataManager() {
        val adapter = mock<BrickRecyclerAdapter>()
        whenever(parent.adapter).thenReturn(adapter)
        whenever(adapter.brickDataManager).thenReturn(mock())

        itemDecoration.getItemOffsets(outRect, view, parent, state)

        assertEquals(0, outRect.left.toLong())
        assertEquals(0, outRect.top.toLong())
        assertEquals(0, outRect.right.toLong())
        assertEquals(0, outRect.bottom.toLong())
    }

    @Test
    fun testInvalidWrongAdapterType() {
        whenever(parent.adapter).thenReturn(mock())

        itemDecoration.getItemOffsets(outRect, view, parent, state)

        assertEquals(0, outRect.left.toLong())
        assertEquals(0, outRect.top.toLong())
        assertEquals(0, outRect.right.toLong())
        assertEquals(0, outRect.bottom.toLong())
    }

    @Test
    fun testAllOuter() {
        whenever(brick.isInFirstRow).thenReturn(true)
        whenever(brick.isInLastRow).thenReturn(true)
        whenever(brick.isOnLeftWall).thenReturn(true)
        whenever(brick.isOnRightWall).thenReturn(true)

        itemDecoration.getItemOffsets(outRect, view, parent, state)

        assertEquals(OUTER_LEFT.toLong(), outRect.left.toLong())
        assertEquals(OUTER_TOP.toLong(), outRect.top.toLong())
        assertEquals(OUTER_RIGHT.toLong(), outRect.right.toLong())
        assertEquals(OUTER_BOTTOM.toLong(), outRect.bottom.toLong())
    }

    @Test
    fun testAllInner() {
        whenever(brick.isInFirstRow).thenReturn(false)
        whenever(brick.isInLastRow).thenReturn(false)
        whenever(brick.isOnLeftWall).thenReturn(false)
        whenever(brick.isOnRightWall).thenReturn(false)

        itemDecoration.getItemOffsets(outRect, view, parent, state)

        assertEquals(INNER_LEFT.toLong(), outRect.left.toLong())
        assertEquals(INNER_TOP.toLong(), outRect.top.toLong())
        assertEquals(INNER_RIGHT.toLong(), outRect.right.toLong())
        assertEquals(INNER_BOTTOM.toLong(), outRect.bottom.toLong())
    }

    companion object {
        private const val INNER_LEFT = 1
        private const val INNER_TOP = 2
        private const val INNER_RIGHT = 3
        private const val INNER_BOTTOM = 4
        private const val OUTER_LEFT = 5
        private const val OUTER_TOP = 6
        private const val OUTER_RIGHT = 7
        private const val OUTER_BOTTOM = 8
    }
}
